h2/src/main/org/h2/engine/Session.java

```diff
@@ -1814,39 +1814,38 @@ private long getStatementSavepoint() {
     public void startStatementWithinTransaction(Command command) {
         Transaction transaction = getTransaction();
         if (transaction != null) {
-            HashSet<MVMap<?, ?>> currentMaps = null, allMaps = null;
+            HashSet<MVMap<?, ?>> maps = new HashSet<>();
             if (command != null) {
                 Set<DbObject> dependencies = command.getDependencies();
-                currentMaps = new HashSet<>();
-                for (DbObject dependency : dependencies) {
-                    if (dependency instanceof MVTable) {
-                        addTableToDependencies((MVTable) dependency, currentMaps);
-                    }
-                }
                 switch (transaction.getIsolationLevel()) {
-                case REPEATABLE_READ: {
-                    allMaps = new HashSet<>();
-                    HashSet<MVTable> processed = new HashSet<>();
-                    for (DbObject dependency : dependencies) {
-                        if (dependency instanceof MVTable) {
-                            addTableToDependencies((MVTable) dependency, allMaps, processed);
-                        }
-                    }
-                    break;
-                }
                 case SNAPSHOT:
                 case SERIALIZABLE:
                     if (!transaction.hasStatementDependencies()) {
-                        allMaps = new HashSet<>();
                         for (Table table : database.getAllTablesAndViews(false)) {
                             if (table instanceof MVTable) {
-                                addTableToDependencies((MVTable) table, allMaps);
+                                addTableToDependencies((MVTable)table, maps);
                             }
                         }
                     }
+                    //$FALL-THROUGH$
```

Comment:
Execution flow shouldn't go below. If this is a first run, there in nothing to collect any more, `break` should be placed inside the `if` condition. If this is a following run, there is no need to collect foreign key dependencies for snapshots at start of the command like `REPEATABLE READ` does, it creates an unnecessary overhead. In that case the following code from `READ_COMMITTED` should be used.

Reply:
`REPEATABLE READ` collects these additional dependencies only for the snapshot at start of the transaction for possible use in the following commands, the current command can't use them, no need to refresh more maps on `SNAPSHOT` / `SERIALIZABLE`.

Of course, they still must be processed before each command on `REPEATABLE READ` level, because entirely new dependencies may appear and they should be appended to the snapshot at start of the transaction in consistent state.

Reply:
While I missed “break” for the initial run of SERIALIZABLE/SNAPSHOT, I think subsequent runs should collect statement level snapshots even for constraint-produced dependencies, or transaction’s own changes won’t be visible.

Reply:
If some table is not used by the current command, it doesn't need a fresh statement-level snapshot of it, it is not going to be used anyway. It can be used later in the following commands, but such commands will have own statement-level snapshots, so an implicit change caused by a constraint can't be missed.

Reply:
Then why are you collecting them for REPEATABLE READ case? Isn’t the same argument applies there?

Reply:
I collect the snapshots at start of the transaction and additions to them with constrains, because otherwise first command may get a snapshot of table A, and the second command may get an additional snapshot of table B and continue to use the old snapshot of table A. These snapshots are used for committed values, but they can be inconsistent, a referenced row may not exist, for example.

The temporary snapshots around the command should not have such additions. They are used for own uncommitted changes only.

The resulting iterator combines committed changes at start of the transaction (or from an addition to it) with uncommitted own changes at start of the command.

Reply:
I current implementation `SERIALIZABLE`, `SNAPSHOT`, and `REPEATABLE READ` use snapshots at start of the transaction (primary map) (with possible following additions on `REPEATABLE READ`) for committed rows and temporary snapshots (secondary map) around the command for own uncommitted changes.

`READ COMMITTED` only uses temporary snapshots around the command for committed and own uncommitted changes, both committed and uncommitted rows are read from the same map. But these snapshots were stored in the same place, where snapshots at start of the transaction are stored (primary map). This allows to use the same implementation of `CommittedIterator` for this and for higher levels when they don't have own (uncommitted) changes, because iteration only over one map is faster than iteration over two maps.

`READ UNCOMMITTED` only uses temporary snapshots around the command for all committed and uncommitted changes. In my implementation it uses the secondary map for them, like on the `SNAPSHOT` level, but it doesn't matter.

In the `Session` the situation is different. `READ COMMITTED` and `READ UNCOMMITTED` use the same variable and parameter for their maps (because they are collected in the same way as dependencies of the current command). `Transaction.markStatementStart()`, however, puts them into one map or into another one depending on isolation level.

Reply:
That’s all understood. With my changes snapshots are always collected as statement-level, and then promoted to transaction level, but only if there is none yet. End result should be the same.
Nevertheless, I think my question is still unanswered - why REPEATABLE READ collects snapshots for  constraint-induced dependencies, and READ COMMITTED does not?

Reply:
Why referential constraint-induced dependencies are ever considered for read snapshots? We are not reading but modifying them. And ref. int. constraints have nothing to do with tx isolation, they should be enforced similar to index uniqueness, where Validation iterator is used regardless.

Reply:
Snapshots around the command don't need additional tables. `READ COMMITTED` uses only them.

`REPEATABLE READ` also has snapshots at start of the transaction (plus additions). These snapshots are never refreshed. If table B has foreign key to table A and the first query uses only the A and the second uses also the B and there were changes in the middle with a new row in A and a new row in B with reference to that row we'll see violation of referential integrity without such addition, because B will have a row and A will not. I don't know whether it is allowed or not, but from my point of view such violations shouldn't exist.

Reply:
I mean committed changes from concurrent transaction. `READ COMMITTED`, `SNAPSHOT`, and `SERIALIZABLE` can't see such violation, why it should exist with `REPEATABLE READ`?

Reply:
But temporary maps on `SERIALIZABLE` for own uncommitted changes don't need such additions, so the execution flow should be adjusted and go to `READ COMMITTED` instead on all runs after the first one. First run doesn't need it at all.

Reply:
>I don't know whether it is allowed or not, but from my point of view such violations shouldn't exist.

Ok, I see your rationale now, although IMHO it should be allowed, as long as it is "REPEATABLE". If you update that record in B and validate referential constraint, you should not be using any snapshots for this validation.

Reply:
I mean that even read-only transaction may discover such unexpected violation due to partial snapshots and it's better to have such additional protection.

Reply:
I also think that it may be allowed, at least I don't see such requirement in the Standard. But it isn't nice and may be surprising for somebody.

Reply:
Agreed, it's better to have such additional protection, especially if cost is not high in terms of performance.
Same goes for snapshots at READ_UNCOMMITTED level.